### ISSUE TITLE
Evaluation

### ISSUE THREAD

**Issue by maximedb** [Kyubyong/transformer#12]
If I understood correctly, at evaluation you run

`preds = np.zeros((hp.batch_size, hp.maxlen), np.int32)`
`for j in range(hp.maxlen):`
`    _preds = sess.run(g.preds, {g.x: x, g.y: preds})`
`    preds[:, j] = _preds[:, j]`

Does that mean that the encoding part runs at every timestep of the decoding process ?

Thanks for the great work 👍

**Comment by madisonmay** [comment]
I'm a little late (and not the original author), but yes that is the correct interpretation.  This is a hack for conceptual simplicity of the evaluation code. The decoder must attend to all previous time steps, but the decoder inputs are not known until we evaluate the softmax of the logits output at the previous timestep, so at the very least we need to run decode again in order to produce the next token. If you poke around at the official implementation in tensor2tensor, you'll see that they make this step more efficient by incrementally caching results during evaluation.
